tensorflow/python/kernel_tests/xent_op_deterministic_test.py

```diff
@@ -94,39 +104,66 @@ def _generateInputs(self, dtype, seed=123):
 
   @test_util.run_in_graph_and_eager_modes
   def testForward(self):
-    with self.session(), test_util.force_cpu():
+    with self.cached_session(), test_util.device(use_gpu=True):
       for dtype in [np.float16, np.float32, np.float64]:
         for trial in range(5):
           seed = 123 + trial
-          labels, logits = self._generateInputs(dtype, seed=seed)
-          result_a = nn_ops.softmax_cross_entropy_with_logits_v2(
+          labels, logits = self._generateInputs(
+              dtype, seed=seed, forward_not_backward=True)
+          result_a = nn_ops.softmax_cross_entropy_with_logits(
               labels=labels, logits=logits)
-          result_b = nn_ops.softmax_cross_entropy_with_logits_v2(
+          result_b = nn_ops.softmax_cross_entropy_with_logits(
               labels=labels, logits=logits)
           self.assertAllEqual(result_a, result_b)
 
   @test_util.run_in_graph_and_eager_modes
   def testBackward(self):
-    with self.session(), test_util.force_cpu():
+    with self.cached_session(), test_util.device(use_gpu=True):
       for dtype in [np.float16, np.float32, np.float64]:
         labels, logits = self._generateInputs(dtype, seed=456)
         output_shape = labels.shape[0]
 
-        def gradients(seed=789):
+        def gradients(seed):
           np.random.seed(seed)
           upstream_gradients = self._randomFloats(output_shape, dtype)
           with backprop.GradientTape(persistent=True) as tape:
+            tape.watch(labels)
             tape.watch(logits)
-            op_output = nn_ops.softmax_cross_entropy_with_logits_v2(
+            op_output = nn_ops.softmax_cross_entropy_with_logits(
                 labels=labels, logits=logits)
             gradient_injector_output = op_output * upstream_gradients
-          return tape.gradient(gradient_injector_output, logits)
+          return tape.gradient(gradient_injector_output, [labels, logits])
 
         for trial in range(5):
           seed = 456 + trial
-          result_a = gradients(seed=seed)
-          result_b = gradients(seed=seed)
-          self.assertAllEqual(result_a, result_b)
+          labels_grad_a, logits_grad_a = gradients(seed=seed)
+          labels_grad_b, logits_grad_b = gradients(seed=seed)
+          self.assertAllEqual(labels_grad_a, labels_grad_b)
+          self.assertAllEqual(logits_grad_a, logits_grad_b)
+
+  # Modifications to the parent class (xent_op_test_base.XentOpTestBase) follow
+
+  def testSingleClass(self):
+    """Disable testing of backprop for single-class case.
+
+    The deterministic implementation does not produce the gradients expected by
+    the test.
+
+    TODO(duncanriach): Identify the source of the difference in gradients for
+                       this case.
+    """
+    self._testSingleClass(test_backprop=False)
```

Comment:
How incorrect are the gradients here and the next test? I am slightly worried there is a correctness problem, although the single class case is seems fairly unimportant.

Reply:
Agreed, it is concerning, and the gradients are very different (if I remember correctly). I was already intending to change these tests so that they do check expected values for both deterministic and nondeterministic cases. The next commit will include that, so we'll have documentation in the test of the difference between the gradients for deterministic vs nondeterministic.

Reply:
After further investigation, it has been revealed that the gradients only mismatch between the nondeterministic and deterministic implementations when the labels vector is not a valid probability distribution as required, but not enforced, by the API. Below are the docstring notes from the `xent_op_deterministic_test.py` in its current state (after the most recent commit).

`testSingleClass`:

```
The deterministic implementation does not produce the gradients expected by
the original test (for the nondeterministic functionality) when the labels
vector is not a valid probability distribution.

labels: [[-1.], [0.], [1.], [1.]]
logits: [[1.], [-1.], [0.], [1.]]

               nondeterministic               deterministic
dloss/dlogits: [[2.0], [1.0], [0.0], [0.0]]   [[0.0], [0.0], [0.0], [0.0]]

Note that only the second two label vectors are a valid probability
distributions (as required by the API) and that the gradient matches for
those cases.
```

`testLabelsBroadcast`:

```
The deterministic implementation does not produce the gradients expected by
the original test (for the nondeterministic functionality) when the labels
vector (after broadcasting) is not a valid probability distribution.

labels: [[0.], [2.], [0.25]]
logits: [[1., 1., 1., 1.],
         [1., 2., 3., 4.],
         [1., 2., 3., 4.]]

dloss/dlogits (nondeterministic):
    [[ 0.25 ,  0.25 ,  0.25 ,  0.25 ],
     [-1.968, -1.913, -1.763, -1.355],
     [-0.218, -0.163, -0.013,  0.394]]

dloss/dlogits (determinsitic):
     [[ 0.   ,  0.   ,  0.   ,  0.   ],
     [-1.743, -1.303, -0.105,  3.150],
     [-0.218, -0.163, -0.013, 0.394]]

Note that neither of the first two broadcast label vectors is a valid
probability distribution (as required by the API) and that these are the
cases that yield different gradients for nondeterministic vs determinsitic
implementations.
```

Reply:
I propose that we remove these illegal label cases from the tests, replacing them with legal cases, thereby making the tests run the same way on both the original implementation and the deterministic implementation.

Reply:
I am slightly worried some people might accidentally rely on the current behavior when given illegal label cases. Also in practice, the labels might not add up exactly to 1 due to rounding errors. I think we should keep it as is for now, with the TODO to investigate this. I'll later try to think if there is a valid interpretation of softmax-cross-entropy when the labels do not add to 1.